fuzz/provider.c

```diff
@@ -0,0 +1,687 @@
+    { \
+        STACK_OF(evp) *digest_stack = stack;  \
+        \
+        if (sk_##evp##_push(digest_stack, digest) > 0) \
+            evp##_up_ref(digest); \
+    } \
+    static void init_##name(OSSL_LIB_CTX * libctx) \
+    { \
+        name##_collection = sk_##evp##_new(cmp_##evp); \
+        evp##_do_all_provided(libctx, collect_##evp, name##_collection); \
+    } \
+    static void cleanup_##name(void) \
+    { \
+        sk_##evp##_pop_free(name##_collection, evp##_free); \
+    }
+
+DEFINE_ALGORITHMS(digests, EVP_MD)
+
+DEFINE_ALGORITHMS(kdf, EVP_KDF)
+
+DEFINE_ALGORITHMS(cipher, EVP_CIPHER)
+
+DEFINE_ALGORITHMS(kem, EVP_KEM)
+
+DEFINE_ALGORITHMS(keyexch, EVP_KEYEXCH)
+
+DEFINE_ALGORITHMS(rand, EVP_RAND)
+
+DEFINE_ALGORITHMS(mac, EVP_MAC)
+
+DEFINE_ALGORITHMS(keymgmt, EVP_KEYMGMT)
+
+DEFINE_ALGORITHMS(signature, EVP_SIGNATURE)
+
+DEFINE_ALGORITHMS(asym_ciphers, EVP_ASYM_CIPHER)
+
+static OSSL_LIB_CTX *libctx = NULL;
+
+int FuzzerInitialize(int *argc, char ***argv)
+{
+    libctx = OSSL_LIB_CTX_new();
+    if (libctx == NULL)
+        return 0;
+
+    init_digests(libctx);
+    init_kdf(libctx);
+    init_cipher(libctx);
+    init_kem(libctx);
+    init_keyexch(libctx);
+    init_rand(libctx);
+    init_mac(libctx);
+    init_keymgmt(libctx);
+    init_signature(libctx);
+    init_asym_ciphers(libctx);
+    return 1;
+}
+
+void FuzzerCleanup(void)
+{
+    cleanup_digests();
+    cleanup_kdf();
+    cleanup_cipher();
+    cleanup_kem();
+    cleanup_keyexch();
+    cleanup_rand();
+    cleanup_mac();
+    cleanup_keymgmt();
+    cleanup_signature();
+    cleanup_asym_ciphers();
+
+    OSSL_LIB_CTX_free(libctx);
+}
+
+static int read_uint(const uint8_t **buf, size_t *len, uint64_t **res)
+{
+    int r = 1;
+
+    if (*len < sizeof(uint64_t)) {
+        r = 0;
+        goto end;
+    }
+
+    *res = malloc(sizeof(uint64_t));
+    **res = (uint64_t) **buf;
+
+    *buf += sizeof(uint64_t);
+    *len -= sizeof(uint64_t);
+end:
+    return r;
+}
+
+static int read_int(const uint8_t **buf, size_t *len, int64_t **res)
+{
+    int r = 1;
+
+    if (*len < sizeof(int64_t)) {
+        r = 0;
+        goto end;
+    }
+
+
+    *res = malloc(sizeof(int64_t));
+    **res = (int64_t) **buf;
+
+    *buf += sizeof(int64_t);
+    *len -= sizeof(int64_t);
+end:
+    return r;
+}
+
+static int read_double(const uint8_t **buf, size_t *len, double **res)
+{
+    int r = 1;
+
+    if (*len < sizeof(double)) {
+        r = 0;
+        goto end;
+    }
+
+    *res = malloc(sizeof(double));
+    **res = (double) **buf;
+
+    *buf += sizeof(double);
+    *len -= sizeof(double);
+end:
+    return r;
+}
+
+static int read_utf8_string(const uint8_t **buf, size_t *len, char **res)
+{
+    int r;
+    size_t i;
+    const uint8_t *ptr = *buf;
+    int found = 0;
+
+    for (i = 0; i < *len; ++i) {
+        if (*ptr == 0) {
+            ptr++;
+            found = 1;
+            break;
+        }
+        ptr++;
+    }
```

Comment:
Can't you just use strnlen?

Reply:
This would be the alterantive implementation:

```
static int read_utf8_string(const uint8_t **buf, size_t *len, char **res)
{
    int r;
    size_t i;
    const uint8_t *ptr;

    size_t end_index = strnlen((const char *) *buf, *len);

    ptr = buf + end_index;

    *res = (char *) *buf;

    r = ptr - *buf;
    *len -= r;
    *buf = ptr;

end:
    return r;
}
```